better_exceptions/highlighter.py

```diff
@@ -0,0 +1,85 @@
+import builtins
+import io
+import keyword
+import tokenize
+
+
+STYLE = {
+    'builtin': '\x1b[35;1m{}\x1b[m',
+    'comment': '\x1b[2;37m{}\x1b[m',
+    'constant': '{}',
+    'identifier': '{}',
+    'keyword': '\x1b[33;1m{}\x1b[m',
+    'number': '\x1b[31m{}\x1b[m',
+    'operator': '{}',
+    'other': '{}',
+    'punctuation': '{}',
+    'string': '\x1b[31m{}\x1b[m',
+}
+
+
+class Highlighter(object):
+
+    def __init__(self, style):
+        self._style = style
+        self._builtins = dir(builtins)
+        self._punctation = {'(', ')', '[', ']', '{', '}', ':', ',', ';'}
+        self._constants = {'True', 'False', 'None'}
+
+    def highlight(self, source):
+        style = self._style
+        row, column = 0, 0
+        output = ""
+
+        for token in self.tokenize(source):
+            type_, string, start, end, line = token
+
+            if type_ == tokenize.NAME:
+                if string in self._constants:
+                    color = style['constant']
+                elif keyword.iskeyword(string):
+                    color = style['keyword']
+                elif string in self._builtins:
+                    color = style['builtin']
+                else:
+                    color = style['identifier']
+            elif type_ == tokenize.OP:
+                if string in self._punctation:
+                    color = style['punctuation']
+                else:
+                    color = style['operator']
+            elif type_ == tokenize.NUMBER:
+                color = style['number']
+            elif type_ == tokenize.STRING:
+                color = style['string']
+            elif type_ == tokenize.COMMENT:
+                color = style['comment']
+            else:
+                color = style['other']
+
```

Comment:
Yes this is why I initially wanted to avoid tokenizers, because of the extra syntax highlighting branches and stuff. But I don't think that's a problem now. :)

Reply:
Oh, I see. I thought it was ok as it's quite easy to comprehend, but I agree that those nested ifs are inelegant. 😕
I don't know how I could improve it unfortunately.

Reply:
Could those token types be put in a map?

Reply:
The tokens could be stored in a dict, but how to handle the second branching:
```python
if string in self._constants:
    color = style['constant']
elif keyword.iskeyword(string):
    color = style['keyword']
elif string in self._builtins:
    color = style['builtin']
else:
    color = style['identifier']
```

Reply:
Hmm true. Yeah, leave it as is. Better to be explicit like this than to try to be 'clever' with it.

LGTM.